crates/subspace-service/src/domains/request_handler.rs

```diff
@@ -0,0 +1,248 @@
+where
+    Block: BlockT,
+    Client: ProvideRuntimeApi<Block>
+        + BlockBackend<Block>
+        + ProofProvider<Block>
+        + HeaderBackend<Block>
+        + Send
+        + Sync
+        + 'static,
+    Client::Api: DomainsApi<Block, DomainHeader> + FraudProofApi<Block, DomainHeader>,
+    DomainHeader: Header,
+{
+    /// Create a new [`LastDomainBlockERRequestHandler`].
+    pub fn new<NB>(
+        protocol_id: &ProtocolId,
+        fork_id: Option<&str>,
+        client: Arc<Client>,
+        num_peer_hint: usize,
+    ) -> (Self, NB::RequestResponseProtocolConfig)
+    where
+        NB: NetworkBackend<Block, <Block as BlockT>::Hash>,
+    {
+        // Reserve enough request slots for one request per peer when we are at the maximum
+        // number of peers.
+        let capacity = std::cmp::max(num_peer_hint, 1);
+        let (tx, request_receiver) = async_channel::bounded(capacity);
+
+        let protocol_config = generate_protocol_config::<_, Block, NB>(
+            protocol_id,
+            client
+                .block_hash(0u32.into())
+                .ok()
+                .flatten()
+                .expect("Genesis block exists; qed"),
+            fork_id,
+            tx,
+        );
+
+        (
+            Self {
+                request_receiver,
+                client,
+                _phantom: PhantomData,
+            },
+            protocol_config,
+        )
+    }
+
+    /// Run [`StateRequestHandler`].
+    pub async fn run(mut self) {
+        while let Some(request) = self.request_receiver.next().await {
+            let IncomingRequest {
+                peer,
+                payload,
+                pending_response,
+            } = request;
+
+            match self.handle_request(payload, pending_response, &peer) {
+                Ok(()) => debug!("Handled domain block info request from {}.", peer),
+                Err(e) => error!(
+                    "Failed to handle domain block info request from {}: {}",
+                    peer, e,
+                ),
+            }
+        }
+    }
+
+    fn handle_request(
+        &mut self,
+        payload: Vec<u8>,
+        pending_response: oneshot::Sender<OutgoingResponse>,
+        peer: &PeerId,
+    ) -> Result<(), HandleRequestError> {
+        let request = LastConfirmedBlockRequest::<Block>::decode(&mut payload.as_slice())?;
+
+        trace!("Handle last confirmed domain block info request: {peer}, request: {request:?}",);
+
+        let result = {
+            let target_block_hash = if let Some(block_hash) = request.block_hash {
+                block_hash
+            } else {
+                let info = self.client.info();
+                info.best_hash
+            };
+
+            let last_confirmed_block_receipt = self
+                .client
+                .runtime_api()
+                .last_confirmed_domain_block_receipt(target_block_hash, request.domain_id);
```

Comment:
Perhaps add runtime api version check before calling into the runtime api, otherwise, it will always return an error if the runtime is not deployed to gemini-3h or the node is not synced yet. See the following as an example:
 https://github.com/autonomys/subspace/blob/129f35f0da17d5721dc42607961aa5647a807cc3/domains/client/domain-operator/src/domain_block_processor.rs#L770-L782